### PR TITLE
Attribute generation

### PR DESCRIPTION
**Pull request by leifj**
This one is "clean"... I think

### CODE REVIEW

File: src/satosa/micro_services/attribute_generation.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+evaluated with a single context that represents the response attributes
+before the microservice is processed. De-referencing the attribute
+name as in '{{name}}' results in a ';'-separated list of all attribute
+values. This notation is useful when you know there is only a single
+attribute value in the set.
+
+*Special contexts*
+
+For treating the values as a list - eg for interating using mustach,
+use the .values sub-context For instance to synthesize all fist-last
```
Comment (johanlundberg) [inline src/satosa/micro_services/attribute_generation.py]: all first-last

File: src/satosa/micro_services/attribute_generation.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
+import re
+import pystache
+
+from .base import ResponseMicroService
+
+def _config(f, requester, provider):
```
Comment (johanlundberg) [inline src/satosa/micro_services/attribute_generation.py]: Do you want to update your code to use the util function added in #99?

File: src/satosa/micro_services/attribute_generation.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    def __str__(self):
+        return ";".join(self._values)
+
+    @property
+    def values(self):
+        [{self._attr_name: v} for v in self._values]
+   
+    @property 
+    def value(self):
+        if 1 == len(self._values):
```
Comment (johanlundberg) [inline src/satosa/micro_services/attribute_generation.py]: Just picky but I really find len(self._values) == 1 is much easier to read.

File: src/satosa/micro_services/attribute_generation.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        syn_attributes = dict()
+        context = dict()
+        
+        for attr_name,values in attributes.items():
+           context[attr_name] = MustachAttrValue(attr_name, values)
+
+        recipes = _config(self.synthetic_attributes, requester, provider)
+        print(context)
+        for attr_name, fmt in recipes:
+           print(fmt)
```
Comment (johanlundberg) [inline src/satosa/micro_services/attribute_generation.py]: Rogue print statement

File: src/satosa/micro_services/attribute_generation.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+           syn_attributes[attr_name] = [v.strip().strip(';') for v in re.split("[;\n]+", pystache.render(fmt, context))]
+        print(syn_attributes)
```
Comment (johanlundberg) [inline src/satosa/micro_services/attribute_generation.py]: Rogue print statement

File: src/satosa/micro_services/attribute_generation.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+    def _synthesize(self, attributes, requester, provider):
+        syn_attributes = dict()
+        context = dict()
+        
+        for attr_name,values in attributes.items():
+           context[attr_name] = MustachAttrValue(attr_name, values)
+
+        recipes = _config(self.synthetic_attributes, requester, provider)
+        print(context)
```
Comment (johanlundberg) [inline src/satosa/micro_services/attribute_generation.py]: Rogue print statement

File: src/satosa/micro_services/attribute_generation.py (hunk 2 above)
Comment (leifj) [inline src/satosa/micro_services/attribute_generation.py]: I was planning to do that in a later refactor - dunno if I can cherry-pick that function

**Comment by leifj** [comment]
pushed fixes for all review comments except the one that requires merging of #99

**Comment by leifj** [comment]
finished the refactor too

**Pull request merged**